File: src-capacitor/note-node/android/src/main/java/io/chainbow/note_node/model/Dtos.kt
package io.chainbow.note_node.model

import com.chibatching.kotpref.KotprefModel
import com.chibatching.kotpref.gsonpref.gsonPref
import io.realm.RealmObject
import io.realm.annotations.PrimaryKey
import org.json.JSONObject

open class ID {
    var _id: String = ""
}

class Seed {
    var _id: String? = null
    var salt: String? = null
    var doc: String? = null

    companion object {
        const val SEED: String = "seed"
    }
}

object Store : KotprefModel() {
    var seed by gsonPref(Seed())
}

open class NoteRecord : RealmObject() {
    @PrimaryKey
    var _id: String = ""
    var index: Long? = null
    var address: String? = null
    var tms: Long? = null
    var noteScriptHex: String? = null
    var decryptedNote: String? = null
    var tx_status: Int? = null
    var tx_height: Long? = null
    var tx_hash: String? = null
    var raw: String? = null
    var draft: Boolean? = null
    var del: Boolean = false
    var sharer: String? = null
    var addressIndex: Long = 0
    var outputIndex: Long = 0

    companion object {
        const val START_WITH: String = "note_"
    }
}

//open class DecryptedNote : RealmObject() {
//    @PrimaryKey
//    var _id: String = ""
//    var ttl: String? = null
//    var fid: String? = null
//    var pwd: String? = null
//    var url: String? = null
//    var otp: String? = null
//    var mem: String? = null
//    var tms: Long? = null
//    var del: Boolean? = null
//}

open class TxRecord : RealmObject() {
    @PrimaryKey
    var _id: String = ""
    var tx_height: Long? = null
    var raw: String? = null
    var income: Long? = null
    var outcome: Long? = null
    var type: String? = null
    var time: Long? = null
    var tx_hash: String? = null
    var address: String? = null
}

open class NoteTxRecord : RealmObject() {
    @PrimaryKey
    var _id: String = ""
    var index: Long? = null
    var address: String? = null
    var tms: Long? = null
    var noteScriptHex: String? = null
    var tx_status: Int? = null
    var tx_height: Long? = null
    var tx_hash: String? = null
    var updated: Long? = null
    var raw: String? = null
    var draft: Boolean? = null
    var outputIndex: Long = 0
    var addressIndex: Long = 0
    var sharer: String? = null
    var del: Boolean = false

    companion object {
        const val START_WITH: String = "tx_"
    }
}

class Finder {
    var selector: Selector? = null
    var fields: Array<String>? = null
    var sort: Array<String>? = null
    var tx_hash: String? = null
}

class Selector {
    var _id: String? = null
    var tx_status: Int? = null
    var draft: Boolean? = null
    var del: Boolean? = null
}

class Note {
    var ttl: String? = null
    var fid: String? = null
    var pwd: String? = null
    var url: String? = null
    var otp: String? = null
    var mem: String? = null
    var tms: Long? = null
}

class NoteStructFront {
    var id: Long? = null
    var note: JSONObject? = null
    var status: NoteStatus? = null
}

class NoteStatus {
    var tx_status: Int? = null
    var tx_height: Long? = null
    var tx_hash: String? = null
}

File: src-capacitor/ios/App/App/model/Dtos.swift
//
// Created by 蘇豪 on 2020/04/04.
//

import Foundation
import RealmSwift
import SwiftyJSON

class ID {
    @objc dynamic var _id: String = ""
}

struct Seed {
    public static let SEED = "seed"

    static let _id = "_id"
    static let salt = "salt"
    static let doc = "doc"

    static func remove() {
        UserDefaults.standard.removeObject(forKey: _id)
        UserDefaults.standard.removeObject(forKey: salt)
        UserDefaults.standard.removeObject(forKey: doc)
    }
}

//笔记对象，只保留最新笔记
class NoteRecord: Object {
    public static let START_WITH = "note_"

    @objc dynamic var _id: String = ""
    @objc dynamic var index: Int64 = 0
    @objc dynamic var address: String? = nil
    @objc dynamic var tms: Int64 = 0
    @objc dynamic var noteScriptHex: String? = nil
    @objc dynamic var decryptedNote: String? = nil
    @objc dynamic var tx_status = 0
    @objc dynamic var tx_height = 0
    @objc dynamic var tx_hash: String? = nil
    @objc dynamic var raw: String? = nil
    @objc dynamic var draft = false
    @objc dynamic var del = false
    @objc dynamic var sharer: String? = nil
    @objc dynamic var addressIndex: Int64  = 0
    @objc dynamic var outputIndex: Int64  = 0

    override static func primaryKey() -> String? {
        "_id"
    }
}

//class DecryptedNote: Object {
//    @objc dynamic var _id: String = ""
//    @objc dynamic var ttl: String? = nil
//    @objc dynamic var fid: String? = nil
//    @objc dynamic var pwd: String? = nil
//    @objc dynamic var url: String? = nil
//    @objc dynamic var otp: String? = nil
//    @objc dynamic var mem: String? = nil
//    @objc dynamic var tms = 0
//    @objc dynamic var del = false
//
//    override static func primaryKey() -> String? {
//        "_id"
//    }
//}

//普通交易记录，来自网络浏览器
class TxRecord: Object {
    @objc dynamic var _id: String = ""
    @objc dynamic var height = 0
    @objc dynamic var raw: String? = nil
    @objc dynamic var income: Int64  = 0
    @objc dynamic var outcome: Int64  = 0
    @objc dynamic var type: String? = nil
    @objc dynamic var time: Int64  = 0
    @objc dynamic var tx_hash: String? = nil
    @objc dynamic var address: String? = nil

    override static func primaryKey() -> String? {
        "_id"
    }
}

class NoteTxRecord: Object {
    public static let START_WITH = "tx_"

    @objc dynamic var _id: String = ""
    @objc dynamic var index: Int64 = 0
    @objc dynamic var address: String? = nil
    @objc dynamic var tms: Int64 = 0
    @objc dynamic var noteScriptHex: String? = nil
    @objc dynamic var tx_status = 0
    @objc dynamic var tx_height = 0
    @objc dynamic var tx_hash: String? = nil
    @objc dynamic var outputIndex: Int64  = 0
    @objc dynamic var addressIndex: Int64  = 0
    @objc dynamic var updated: Int64  = 0
    @objc dynamic var raw: String? = nil
    @objc dynamic var sharer: String? = nil
    @objc dynamic var draft = false
    @objc dynamic var del = false

    override static func primaryKey() -> String? {
        "_id"
    }
}
